File: src/lib/scala/Slate.Integration/src/main/scala/slate/integration/FilesApi.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.integration

import slate.common.{Doc, Result, Strings, Files}
import slate.common.Files._
import slate.core.apis.{ApiAction, Api}
import slate.core.cloud.CloudFilesBase
import slate.core.common.svcs.ApiWithSupport


@Api(area = "infra", name = "files", desc = "api info about the application and host", roles= "admin", auth="key-roles", verb = "post", protocol = "*")
class FilesApi(val files:CloudFilesBase) extends ApiWithSupport {


  @ApiAction(name = "", desc= "connect to the datasource", roles= "@parent", verb = "@parent", protocol = "@parent")
  def connectWith(key:String, password:String, tag:String):Unit =
  {
    files.connectWith(key, password, tag)
  }


  @ApiAction(name = "", desc= "creates the root folder/bucket",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def createRootFolder(rootFolder:String):Unit = {
    files.createRootFolder(rootFolder)
  }


  @ApiAction(name = "", desc= "creates a file with the supplied folder name, file name, and content",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def create(folder:String, name:String, content:String):Unit = {
    files.create(folder, name, content)
  }


  @ApiAction(name = "", desc= "creates a file with the supplied folder name, file name, and content from file path",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def createFromPath(folder:String, name:String, filePath:String): Result[String] = {
    files.createFromPath(folder, name, interpretUri(filePath).getOrElse(filePath))
  }


  @ApiAction(name = "", desc= "creates a file with the supplied folder name, file name, and content from doc",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def createFromDoc(folder:String, name:String, doc:Doc): Result[String] = {
    files.create(folder, name, doc.content)
  }


  @ApiAction(name = "", desc= "updates a file with the supplied folder name, file name, and content",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def update(folder:String, name:String, content:String): Result[String] = {
    files.update(folder, name, content)
  }


  @ApiAction(name = "", desc= "updates a file with the supplied folder name, file name, and content from file path",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def updateFromPath(folder:String, name:String, filePath:String): Result[String] =
  {
    files.updateFromPath(folder, name, interpretUri(filePath).getOrElse(filePath))
  }


  @ApiAction(name = "", desc= "updates a file with the supplied folder name, file name, and content from doc",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def updateFromDoc(folder:String, name:String, doc:Doc): Result[String] =
  {
    files.updateFromPath(folder, name, doc.content)
  }


  @ApiAction(name = "", desc= "deletes a file with the supplied folder name, file name",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def delete(folder:String, name:String):Result[String] = {
    files.delete(folder, name)
  }


  @ApiAction(name = "", desc= "downloads the file specified by folder and name to the local folder specified.",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def download(folder:String, name:String, localFolder:String, display:Boolean): Result[String] = {
    show(files.download(folder, name, interpretUri(localFolder).getOrElse(localFolder)), display)
  }


  @ApiAction(name = "", desc= "downloads the file specified by folder and name, as text content to file supplied",
    roles= "@parent", verb = "@parent", protocol = "@parent")
  def downloadToFile(folder:String, name:String, filePath:String, display:Boolean):Result[String] = {
    show(files.downloadToFile(folder, name, interpretUri(filePath).getOrElse(filePath)), display)
  }


  private def show(result:Result[String], display:Boolean): Result[String]  = {
    val path = result.getOrElse("")
    val output = if(display){
      val text = Files.readAllText(path)
      "PATH   : " + path + Strings.newline() +
      "CONTENT: " + text
    }
    else
      "PATH   : " + path
    successOrError[String](result.success, output, result.msg, result.tag, result.ref)
  }
}

File: src/lib/kotlin/slatekit-integration/src/main/kotlin/slatekit/integration/apis/FilesApi.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.integration.apis

import slatekit.apis.Api
import slatekit.apis.Action
import slatekit.apis.AuthModes
import slatekit.apis.Verbs
import slatekit.common.Sources
import slatekit.common.content.Doc
import slatekit.common.encrypt.Encryptor
import slatekit.common.log.Logger
import slatekit.results.Failure
import slatekit.results.Success
import slatekit.results.Try
import slatekit.results.getOrElse

@Api(area = "cloud", name = "files", desc = "api info about the application and host",
        auth = AuthModes.Keyed, roles = ["admin"], verb = Verbs.Auto, sources = [Sources.All])
class FilesApi(val files: slatekit.core.cloud.CloudFiles, override val context: slatekit.common.Context) : slatekit.apis.support.FileSupport {

    override val encryptor: Encryptor? = context.enc
    override val logger: Logger? = context.logs.getLogger()

    @Action(desc = "creates the root folder/bucket")
    fun createRootFolder(rootFolder: String) {
        return files.createRootFolder(rootFolder)
    }

    @Action(desc = "creates a file with the supplied folder name, file name, and content")
    fun create(folder: String, name: String, content: String) {
        files.create(folder, name, content)
    }

    @Action(desc = "creates a file with the supplied folder name, file name, and content from file path")
    fun createFromPath(folder: String, name: String, filePath: String): Try<String> {
        return files.createFromPath(folder, name, slatekit.common.Uris.interpret(filePath) ?: filePath)
    }

    @Action(desc = "creates a file with the supplied folder name, file name, and content from doc")
    fun createFromDoc(folder: String, name: String, doc: Doc): Try<String> {
        return files.create(folder, name, doc.content)
    }

    @Action(desc = "updates a file with the supplied folder name, file name, and content")
    fun update(folder: String, name: String, content: String): Try<String> {
        return files.update(folder, name, content)
    }

    @Action(desc = "updates a file with the supplied folder name, file name, and content from file path")
    fun updateFromPath(folder: String, name: String, filePath: String): Try<String> {
        return files.updateFromPath(folder, name, interpretUri(filePath) ?: filePath)
    }

    @Action(desc = "updates a file with the supplied folder name, file name, and content from doc")
    fun updateFromDoc(folder: String, name: String, doc: Doc): Try<String> {
        return files.updateFromPath(folder, name, doc.content)
    }

    @Action(desc = "deletes a file with the supplied folder name, file name")
    fun delete(folder: String, name: String): Try<String> {
        return files.delete(folder, name)
    }

    @Action(desc = "get file as text")
    fun getAsText(folder: String, name: String):Try<String> {
        return files.getAsText(folder, name)
    }

    @Action(desc = "downloads the file specified by folder and name to the local folder specified.")
    fun download(folder: String, name: String, localFolder: String, display: Boolean): Try<String> {
        return show(files.download(folder, name, interpretUri(localFolder) ?: localFolder), display)
    }

    @Action(desc = "downloads the file specified by folder and name, as text content to file supplied")
    fun downloadToFile(folder: String, name: String, filePath: String, display: Boolean): Try<String> {
        return show(files.downloadToFile(folder, name, slatekit.common.Uris.interpret(filePath) ?: filePath), display)
    }

    private fun show(result: Try<String>, display: Boolean): Try<String> {
        val path = result.getOrElse { "" }
        val output = if (display) {
            val text = java.io.File(path).readText()
            "PATH   : " + path + slatekit.common.newline +
                    "CONTENT: " + text
        } else
            "PATH   : " + path
        return result.fold( { Success(output) }, { Failure(result.msg) }).toTry()
    }
}
